File: platforms/android/contribute/src/main/java/pulse/contribute/android/ApiMap.kt
package pulse.contribute.android

import android.webkit.JavascriptInterface
import org.json.JSONObject

import pulse.contribute.android.api.vfs.File as VFS_File
import pulse.contribute.android.api.vfs.Directory as VFS_Directory
import pulse.contribute.android.api.net.Web as NET_Web
import pulse.contribute.android.api.internal.Native as Internal_Native

class ApiMap() {

    companion object{

        //gets defined internally on App Start

        //Platform api function map for this platform
        fun platformApiMap(): String {
            return JSONObject(hashMapOf(

                "NATIVE_TOAST" to "showNativeToast",
                "NATIVE_NOTIFICATION" to "showNativeNotification",
                "NATIVE_NOTIFICATION_SOUND" to "playNotificationSound",

                "WRITE_FILE" to "writeFile",
                "APPEND_FILE" to "appendFile",
                "READ_FILE" to "readFile",
                "DELETE_FILE" to "deleteFile",
                "EXISTS_FILE" to "existsFile",

                "EXISTS_DIRECTORY" to "existsDirectory",
                "CREATE_DIRECTORY" to "createDirectory",
                "DELETE_DIRECTORY" to "deleteDirectory",
                "READ_DIRECTORY" to "readDirectory",

                "FETCH" to "asyncFetch",
                "DOWNLOAD_FILE" to "downloadFile"

            ) as Map<String, String>).toString()
        }

    }

    //<editor-fold desc="Internal.Native">

    @JavascriptInterface
    fun showNativeToast(promiseId: String, toast: String) = Internal_Native.showNativeToast(promiseId, toast)

    @JavascriptInterface
    fun showNativeNotification(promiseId: String, title: String, message: String, time: Int) = Internal_Native.showNativeNotification(promiseId, title, message, time)

    @JavascriptInterface
    fun playNotificationSound(promiseId: String) = Internal_Native.playNotificationSound(promiseId)

    //</editor-fold>

    //<editor-fold desc="VFS.File">

    @JavascriptInterface
    fun writeFile(promiseId: String, path: String, content: String) = VFS_File.writeFile(promiseId, path, content)

    @JavascriptInterface
    fun appendFile(promiseId: String, path: String, content: String) = VFS_File.appendFile(promiseId, path, content)

    @JavascriptInterface
    fun readFile(promiseId: String, path: String) = VFS_File.readFile(promiseId, path)

    @JavascriptInterface
    fun deleteFile(promiseId: String, path: String) = VFS_File.deleteFile(promiseId, path)

    @JavascriptInterface
    fun existsFile(promiseId: String, path: String) = VFS_File.existsFile(promiseId, path)

    //</editor-fold>

    //<editor-fold desc="VFS.Directory">

    @JavascriptInterface
    fun existsDirectory(promiseId: String, path: String) = VFS_Directory.existsDirectory(promiseId, path)

    @JavascriptInterface
    fun createDirectory(promiseId: String, path: String) = VFS_Directory.createDirectory(promiseId, path)

    @JavascriptInterface
    fun deleteDirectory(promiseId: String, path: String) = VFS_Directory.deleteDirectory(promiseId, path)

    @JavascriptInterface
    fun readDirectory(promiseId: String, path: String) = VFS_Directory.readDirectory(promiseId, path)

    //</editor-fold>

    //<editor-fold desc="NET.Web">

    @JavascriptInterface
    fun asyncFetch(promiseId: String, url: String, optionsJson: String) = NET_Web.asyncFetch(promiseId, url, optionsJson)

    @JavascriptInterface
    fun downloadFile(promiseId: String, url: String, saveTo: String, optionsJson: String, progressEventId: String) = NET_Web.downloadFile(promiseId, url, saveTo, optionsJson, progressEventId)

    //</editor-fold>

}

File: platforms/dot-net.winforms/PulseContribute.DotNet.WinForms/ApiMap.cs
using System.Runtime.InteropServices;

namespace PulseContribute.DotNet.WinForms
{
    [ComVisible(true)]
    public class ApiMap
    {
        [ComVisible(false)]
        public static String PlatformApiMap()
        {
            return JSON.Stringify(new Dictionary<string, string>
            {
                { "WRITE_FILE", "WriteFile" },
                { "APPEND_FILE", "AppendFile" },
                { "READ_FILE", "ReadFile" },
                { "DELETE_FILE", "DeleteFile" },
                { "EXISTS_FILE", "ExistsFile" },

                { "EXISTS_DIRECTORY", "ExistsDirectory" },
                { "CREATE_DIRECTORY", "CreateDirectory" },
                { "DELETE_DIRECTORY", "DeleteDirectory" },
                { "READ_DIRECTORY", "ReadDirectory" },

                { "FETCH", "AsyncFetch" }
            });
        }

        #region VFS.File

        public void WriteFile(string promiseId, string path, string content){ VFS.File.WriteFile(promiseId, path, content); }

        public void AppendFile(string promiseId, string path, string content) { VFS.File.AppendFile(promiseId, path, content); }

        public void ReadFile(string promiseId, string path) { VFS.File.ReadFile(promiseId, path); }

        public void DeleteFile(string promiseId, string path) { VFS.File.DeleteFile(promiseId, path); }

        public void ExistsFile(string promiseId, string path) { VFS.File.ExistsFile(promiseId, path); }

        #endregion

        #region VFS.Directory

        public void ExistsDirectory(string promiseId, string path){ VFS.Directory.ExistsDirectory(promiseId, path); }

        public void CreateDirectory(string promiseId, string path) { VFS.Directory.CreateDirectory(promiseId, path); }

        public void DeleteDirectory(string promiseId, string path) { VFS.Directory.DeleteDirectory(promiseId, path); }

        public void ReadDirectory(string promiseId, string path) { VFS.Directory.ReadDirectory(promiseId, path); }

        #endregion

        #region NET.Web

        public void AsyncFetch(string promiseId, string url, string options) { Net.Web.AsyncFetch(promiseId, url, options); }

        #endregion
    }
}
